classes/local/store/object_file_system.php

```diff
@@ -736,8 +736,39 @@ protected function copy_from_external_to_local($contenthash) {
      * @throws \dml_exception
      */
     public function redirect_to_presigned_url($contenthash, $headers = array()) {
+        global $FULLME;
         try {
-            redirect($this->externalclient->generate_presigned_url($contenthash, $headers));
+            $signedurl = $this->externalclient->generate_presigned_url($contenthash, $headers);
+            if (headers_sent()) {
+                debugging('objectfs redirect for ' . $contenthash . ' from ' . $FULLME .
+                        ': headers already sent; redirect may be incorrectly cached in browser');
+            } else {
+                // Remove all previously-set headers, and look for cache-control setting.
+                $cachecontrol = '';
+                foreach (headers_list() as $header) {
+                    // Get header name (text before the colon).
+                    if (preg_match('~^([^:]+):(.*)$~', $header, $matches)) {
+                        if (strtolower($matches[1]) === 'cache-control') {
+                            $cachecontrol = $matches[2];
+                        }
+                        header_remove($matches[1]);
```

Comment:
It would be nicer if `$matches[1]` and `$matches[2]` were named variables which describes what they are, e.g. `name` and `value` (or `headername` and `headervalue`).

This would likely make it quicker to read and reason about.